File: frontend/src/api/chain.ts
// frontend/src/api/chain.ts
import { apiFetch } from './client'

export interface ChainFactor {
  name: string
  points: number
  max: number
  detail: string
}

export interface ChainCandidate {
  strike: number
  delta: number
  last: number
  bid: number
  ask: number
  spread_pct: number
  volume: number
  open_interest: number
  arr_pct: number
  breakeven: number
  downside_cushion_pct: number
  capital_required: number
  score: number
  factors: ChainFactor[]
}

export interface ChainExpiry {
  expiration_date: string
  dte: number
  day_of_week: string
  earnings_in_window: boolean
  earnings_date: string | null
  candidates: ChainCandidate[]
}

export interface SpTimingSignalSummary {
  ticker: string
  score: number
  grade: string
  iv_percentile: number | null
  atm_iv: number | null
  spot_price: number | null
  factors: ChainFactor[]
  commentary: string | null
  strike_hint: string | null
  caution: string | null
  cached_at: string
  fetch_status: string
  fetch_error: string | null
}

export interface ChainScreenResult {
  ticker: string
  spot: number | null
  strategy: string
  sp_timing_signal: SpTimingSignalSummary | null
  iv_percentile: number | null
  fetched_at: string
  expiries: ChainExpiry[]
}

export interface ChainScreenParams {
  strategy: string
  num_expiries?: number
  candidates_per_expiry?: number
  target_delta?: number
  delta_tolerance?: number
  min_arr_pct?: number
  min_volume?: number
  min_oi?: number
}

export async function screenChain(ticker: string, params: ChainScreenParams): Promise<ChainScreenResult> {
  const query = new URLSearchParams()
  query.set('strategy', params.strategy)
  if (params.num_expiries != null) query.set('num_expiries', String(params.num_expiries))
  if (params.candidates_per_expiry != null) query.set('candidates_per_expiry', String(params.candidates_per_expiry))
  if (params.target_delta != null) query.set('target_delta', String(params.target_delta))
  if (params.delta_tolerance != null) query.set('delta_tolerance', String(params.delta_tolerance))
  if (params.min_arr_pct != null) query.set('min_arr_pct', String(params.min_arr_pct))
  if (params.min_volume != null) query.set('min_volume', String(params.min_volume))
  if (params.min_oi != null) query.set('min_oi', String(params.min_oi))
  return apiFetch<ChainScreenResult>(`/chain/${encodeURIComponent(ticker)}?${query.toString()}`)
}

File: backend/app/schemas/chain.py
# backend/app/schemas/chain.py
from typing import Optional
from pydantic import BaseModel


class ChainFactor(BaseModel):
    name: str
    points: float
    max: float
    detail: str


class ChainCandidate(BaseModel):
    strike: float
    delta: float
    last: float
    bid: float
    ask: float
    spread_pct: float
    volume: int
    open_interest: int
    arr_pct: float
    breakeven: float
    downside_cushion_pct: float
    capital_required: float
    score: float
    factors: list[ChainFactor]


class ChainExpiry(BaseModel):
    expiration_date: str
    dte: int
    day_of_week: str
    earnings_in_window: bool
    earnings_date: Optional[str] = None
    candidates: list[ChainCandidate]


class SpTimingSignalSummary(BaseModel):
    ticker: str
    score: float
    grade: str
    iv_percentile: Optional[float] = None
    atm_iv: Optional[float] = None
    spot_price: Optional[float] = None
    factors: list[dict] = []
    commentary: Optional[str] = None
    strike_hint: Optional[str] = None
    caution: Optional[str] = None
    cached_at: str
    fetch_status: str
    fetch_error: Optional[str] = None


class ChainScreenResponse(BaseModel):
    ticker: str
    spot: Optional[float] = None
    strategy: str
    sp_timing_signal: Optional[SpTimingSignalSummary] = None
    iv_percentile: Optional[float] = None
    fetched_at: str
    expiries: list[ChainExpiry]
